pitmutationmate/src/main/kotlin/com/amos/pitmutationmate/pitmutationmate/execution/ExecutionDoneProcessListener.kt

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+// SPDX-FileCopyrightText: 2023-2024
+
+package com.amos.pitmutationmate.pitmutationmate.execution
+
+import com.amos.pitmutationmate.pitmutationmate.editor.PluginState
+import com.amos.pitmutationmate.pitmutationmate.services.MutationResultService
+import com.amos.pitmutationmate.pitmutationmate.services.RunArchiveService
+import com.amos.pitmutationmate.pitmutationmate.services.UdpMessagingServer
+import com.amos.pitmutationmate.pitmutationmate.ui.ToolWindowFactory
+import com.intellij.codeInsight.daemon.DaemonCodeAnalyzer
+import com.intellij.execution.process.ProcessAdapter
+import com.intellij.execution.process.ProcessEvent
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.wm.ToolWindow
+import com.intellij.openapi.wm.ToolWindowManager
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiFile
+import com.intellij.psi.search.GlobalSearchScope
+
+class ExecutionDoneProcessListener(val project: Project, private val classFqdns: String?) : ProcessAdapter() {
+    private val log: Logger = Logger.getInstance(ExecutionDoneProcessListener::class.java)
+
+    override fun processTerminated(event: ProcessEvent) {
+        log.debug("BasePitestExecutor: executeTask: processTerminated: event: $event")
+        project.service<UdpMessagingServer>().stopServer() // Stop the UDP server
+        if (classFqdns.isNullOrEmpty()) {
+            return
+        }
+
+        // save and get latest pitest results
+        val resultData = project.service<MutationResultService>().updateLastMutationResult()
+        // update tool window with latest result data
+        val toolWindowManager = ToolWindowManager.getInstance(project)
+        val toolWindow: ToolWindow? = toolWindowManager.getToolWindow(ToolWindowFactory.ID)
+        if (toolWindow != null) {
+            toolWindowManager.invokeLater {
+                ToolWindowFactory.Util.updateReport(toolWindow, resultData)
+            }
+        }
+        // restart code highlighting upon new pitest results
+        ApplicationManager.getApplication().runReadAction {
+            restartCodeHighlighting()
+        }
+        // archive the pitest run
+        project.service<RunArchiveService>().archiveRun()
+    }
+
+    private fun restartCodeHighlighting() {
+        PluginState.isAnnotatorEnabled = true
+        val analyser = DaemonCodeAnalyzer.getInstance(project)
+        // moved tool window update since if it's done pitest didn't run through
+        val psiFiles = classFqdns!!.split(",").map { getPsiFileFromFQN(project, it) }
+        psiFiles.forEach { file ->
+            if (file != null) {
+                analyser.restart(file)
```

Comment:
And here, you're ensuring that the CodeAnalyzer is only run on the files that were tested by Pitest, right? So, the CodeAnalyzer no longer runs on the entire project? But couldn't this be slower because now you're searching the whole project multiple times if there are multiple classes in the Pitest run?

Reply:
Hmm could be that JavaPsiFacade.findClass takes a few ms more but as we have it now, every time the annotator runs, the mutation result service also parses the xmls so cutting down the number of annotator runs should reduce overall computing time.
And since the psi information is indexed, I dont think the search takes too long...

Reply:
Ok, that makes sense. But shouldn't we consider changing this in general, with the XML parsing being executed every time we need the results? Do you think this would be possible, at least in the annotator?

Reply:
Definitely! I've already started working on that but didn't want to put more in this PR